File: NeetCode150/stack/2_min_stack.cc
// Time: O(1)
// Space: O(n)
// Algo intution:
//      - maintain two stacks, minStack and normal stack
//      - always push the min val to minStack
//      - pop from both min and normal stack
//      - for top, return stack.top()
//      - for getMin, return minStack.top()
class MinStack {
    private:
        std::stack<int> stack;
        std::stack<int> minStack;
    public:
        MinStack() {
            
        }
        
        void push(int val) {
            stack.push(val);
            val = std::min(val, minStack.empty() ? val : minStack.top());
            minStack.push(val);
        }
        
        void pop() {
            stack.pop();
            minStack.pop();
        }
        
        int top() {
            return stack.top();
        }
        
        int getMin() {
            return minStack.top();
        }
    };
    
    /**
     * Your MinStack object will be instantiated and called as such:
     * MinStack* obj = new MinStack();
     * obj->push(val);
     * obj->pop();
     * int param_3 = obj->top();
     * int param_4 = obj->getMin();
     */

File: NeetCode150/stack/2_min_stack.py
class MinStack:

    def __init__(self):
        self.stack = []
        self.monotonicStack = (
            []
        )  # only adding elements in non-increasing order, skip any elements that break this condition

    def push(self, val: int) -> None:
        self.stack.append(val)

        if self.monotonicStack:
            if val <= self.monotonicStack[-1]:
                self.monotonicStack.append(val)
        else:
            self.monotonicStack.append(val)

    def pop(self) -> None:
        if self.monotonicStack and self.stack[-1] == self.monotonicStack[-1]:
            self.monotonicStack.pop()

        self.stack.pop()

    def top(self) -> int:
        return self.stack[-1]

    def getMin(self) -> int:
        return self.monotonicStack[-1]


# Your MinStack object will be instantiated and called as such:
# obj = MinStack()
# obj.push(val)
# obj.pop()
# param_3 = obj.top()
# param_4 = obj.getMin()

# stack = [-2, 0
# monotonicStack = [-2
# -> -3,
# -> 0
# -> -2
